File: backend/app/api/schemas/timeline.py
"""Schemas for the deployment-timeline insight endpoint."""

from datetime import date

from pydantic import BaseModel


class TrapNightInterval(BaseModel):
    start: date
    end: date
    trap_nights: int


class TimelineDeployment(BaseModel):
    deployment_id: str
    deployment_label: str
    camera_model: str | None
    configured_start: date
    configured_end: date | None
    intervals: list[TrapNightInterval]
    file_count: int


class TimelineSite(BaseModel):
    site_id: str | None
    site_name: str
    deployments: list[TimelineDeployment]


class ConcurrentPoint(BaseModel):
    date: date
    count: int


class HeatmapPoint(BaseModel):
    """One site's media-file count for one calendar day.

    Only days with at least one file get a point, so the payload is bounded
    by (sites x days with files). `site_id` is None for the "(no site)" row,
    matching `TimelineSite.site_id`.
    """

    site_id: str | None
    date: date
    count: int


class TimelineMetrics(BaseModel):
    site_count: int
    deployment_count: int
    total_trap_nights: int
    median_deployment_length_days: float | None
    max_concurrent_cameras: int


class TimelineResponse(BaseModel):
    sites: list[TimelineSite]
    concurrent_cameras: list[ConcurrentPoint]
    heatmap: list[HeatmapPoint]
    metrics: TimelineMetrics
    date_range_from: date | None
    date_range_to: date | None

File: frontend/src/api/timeline.ts
/**
 * Deployment timeline endpoint (Insights → Deployment timeline page).
 *
 * Types mirror backend/app/api/schemas/timeline.py.
 */

import { api } from "../lib/api-client";

export interface TrapNightInterval {
  start: string; // YYYY-MM-DD
  end: string;
  trap_nights: number;
}

export interface TimelineDeployment {
  deployment_id: string;
  deployment_label: string;
  camera_model: string | null;
  configured_start: string;
  configured_end: string | null;
  intervals: TrapNightInterval[];
  file_count: number;
}

export interface TimelineSite {
  site_id: string | null;
  site_name: string;
  deployments: TimelineDeployment[];
}

export interface ConcurrentPoint {
  date: string;
  count: number;
}

/** One site's media-file count for one day. `site_id` is null for "(no site)". */
export interface HeatmapPoint {
  site_id: string | null;
  date: string; // YYYY-MM-DD
  count: number;
}

export interface TimelineMetrics {
  site_count: number;
  deployment_count: number;
  total_trap_nights: number;
  median_deployment_length_days: number | null;
  max_concurrent_cameras: number;
}

export interface TimelineResponse {
  sites: TimelineSite[];
  concurrent_cameras: ConcurrentPoint[];
  heatmap: HeatmapPoint[];
  metrics: TimelineMetrics;
  date_range_from: string | null;
  date_range_to: string | null;
}

export interface TimelineFilters {
  siteIds?: string[];
  dateFrom?: string;
  dateTo?: string;
}

export const timelineApi = {
  get: (projectId: string, filters: TimelineFilters = {}) => {
    const params = new URLSearchParams();
    params.set("project_id", projectId);
    if (filters.siteIds?.length) params.set("site_ids", filters.siteIds.join(","));
    if (filters.dateFrom) params.set("date_from", filters.dateFrom);
    if (filters.dateTo) params.set("date_to", filters.dateTo);
    return api.get<TimelineResponse>(
      `/api/statistics/timeline?${params.toString()}`
    );
  },
};
